### ISSUE TITLE
Generic zipWith

### ISSUE THREAD

**Issue by jhenahan** [haskell-foundation/foundation#134]
As proposed in [this blog post](http://paczesiowa.blogspot.com/2010/03/generalized-zipwithn.html), `zipWith` from `Foundation.Collection` could be generalized. Total blue sky, at this point, so I have no idea what the performance implications are, or if this is too black magic for a core collections class. There's also [Oleg's version](http://okmij.org/ftp/Haskell/polyvariadic.html#ZipN).

I also can't say I grok type families enough to know if there's a really easy encoding of this idea leveraging those.

On one hand, it's very magical. On the other hand, it could be a smaller, more opaque API for this chunk of collections logic, which seems like it would fit the Foundation philosophy.

I'll probably look into exploring some of this over the weekend, but I thought a bit of input from active contributors would be nice to have.

**Comment by vincenthz** [comment]
I can't say I've read the blog post carefully, but it would be useful to define what you're trying to achieve here; From what's the current state, to what's this is trying to achieve. It doesn't have to be working haskell code, just pseudo haskell code or just some signatures.

**Comment by mgsloan** [comment]
I don't think we should generalize zipWith in this way.  Polyvariadic functions in Haskell means you tend to sacrifice polymorphism in the arguments, and get much worse type errors.  There is a reason they aren't idiomatic.

Instead, I would like to see foundation add the `These` datatype (`data These a b = This a | That b | These a b`).  This allows nice definitions for alignable functors.  I'm not sure if we want the Crosswalk stuff, maybe!  But certainly the typeclasses for `Align` and `Unalign` are something to consider.  The following changes should be made to the `Align` class:

* Remove `nil`, it prevents writing instances of `Align` for many types that could have it
* Remove `align`, give primacy to `alignWith`.  Implement `align` as function in terms of `alignWith`.

`Alignable` allows items to be missing on either side of the alignment.   I think this is a more principled zippable.  I'd almost want to define `zipWith` in terms of `alignWith`, however this would be less efficient when one collection is dramatically smaller than the other.

We *should* however, give the following law relating `zipWith` and `alignWith` + `catMaybes`:

```haskell
catMaybes (alignWith (fmap (uncurry f) . justThese)) a b == zipWith f a b

-- Provided by "these" package
justThese (These x y) = Just (x, y)
justThese _ = Nothing
```
